core/sequencer/server_test.go

```diff
@@ -229,3 +252,68 @@ func TestHighWatermarks(t *testing.T) {
 		})
 	}
 }
+
+func TestDuplicateUpdates(t *testing.T) {
+	ctx := context.Background()
+	initMetrics.Do(func() { createMetrics(monitoring.InertMetricFactory{}) })
+	ks, err := keyset.NewHandle(signature.ECDSAP256KeyTemplate())
+	if err != nil {
+		t.Fatalf("signature.NewHandle(): %v", err)
+	}
+	signer, err := signature.NewSigner(ks)
+	if err != nil {
+		t.Fatalf("signature.NewSigner(): %v", err)
+	}
+	authorizedKeys, err := ks.Public()
+	if err != nil {
+		t.Fatalf("Failed to setup tink keyset: %v", err)
+	}
+
+	index := []byte("index")
+	userID := "userID"
+	log0 := []mutator.LogMessage{}
+	mapRev := int64(0)
+	for i, data := range []string{"data1", "data2"} {
+		m := entry.NewMutation(index, directoryID, userID)
+		if err := m.SetCommitment([]byte(data)); err != nil {
+			t.Fatalf("%v", err)
+		}
+		if err := m.ReplaceAuthorizedKeys(authorizedKeys.Keyset()); err != nil {
+			t.Fatalf("%v", err)
+		}
+		update, err := m.SerializeAndSign([]tink.Signer{signer})
+		if err != nil {
+			t.Fatalf("SerializeAndSign(): %v", err)
+		}
+		log0 = append(log0, mutator.LogMessage{
+			ID:        int64(i),
+			Mutation:  update.Mutation,
+			ExtraData: update.Committed},
+		)
+	}
+
+	s := Server{
+		logs: fakeLogs{0: log0},
+		batcher: &fakeBatcher{
+			highestRev: mapRev,
+			batches: map[int64]*spb.MapMetadata{
+				1: {Sources: []*spb.MapMetadata_SourceSlice{
+					{LogId: 0, HighestExclusive: 2},
+				}},
+			},
+		},
+		trillian: &fakeTrillianFactory{
+			tmap: &fakeMap{
+				MapClient:     MapClient{&tclient.MapClient{Conn: &fakeMapConn{}}},
+				latestMapRoot: &types.MapRootV1{Revision: uint64(mapRev)},
+			},
+		},
+	}
+
+	if _, err := s.ApplyRevision(ctx, &spb.ApplyRevisionRequest{
+		DirectoryId: directoryID,
+		Revision:    1,
+	}); status.Code(err) != codes.Unimplemented {
```

Comment:
How did this fail previously?

Reply:
The previous failure was from the trillian map client complaining that there were multiple keys returned for a single index.

Reply:
Added better error checking